rewriter/src/main/java/com/facebook/coresql/rewriter/ApproxPercentileRewriter.java

```diff
@@ -0,0 +1,182 @@
+        String firstArg = unparse(getNthArgument(node, 0).get()).trim();
+        String secondArg = unparse(getNthArgument(node, 1).get()).trim();
+        Formatter formatter = new Formatter(stringBuilder);
+        formatter.format(REPLACEMENT_FORMAT, firstArg, firstArgToPercentiles.get(firstArg), getIndexOfPercentile(firstArg, secondArg));
+        // Lastly, move to end of this node -- we've already added a rewritten version of it to the result, so we don't need to process it further
+        moveToEndOfNode(node);
+    }
+
+    private int getIndexOfPercentile(String firstArg, String secondArg)
+    {
+        return firstArgToPercentiles.get(firstArg).indexOf(secondArg) + ZERO_INDEXING_OFFSET;
+    }
+
+    private static Optional<AstNode> getNthArgument(AstNode node, int n)
+    {
+        Optional<AstNode> argList = Optional.ofNullable(node.GetFirstChildOfKind(JJTARGUMENTLIST));
+        if (!argList.isPresent() || argList.get().jjtGetNumChildren() < n) {
+            return Optional.empty();
+        }
+        AstNode nthArg = (AstNode) argList.get().jjtGetChild(n);
+        return Optional.of(nthArg);
+    }
+
+    private static boolean hasUnsignedLiteralSecondArg(AstNode node)
+    {
+        return getNthArgument(node, 1).filter(astNode -> astNode.getId() == JJTUNSIGNEDNUMERICLITERAL).isPresent();
+    }
+
+    private static boolean isApproxPercentileNode(AstNode node)
+    {
+        Optional<AstNode> identifier = Optional.ofNullable(node.GetFirstChildOfKind(JJTIDENTIFIER));
+        if (!identifier.isPresent()) {
+            return false;
+        }
+        Optional<String> image = Optional.ofNullable(identifier.get().GetImage());
+        return image.isPresent() && image.get().equalsIgnoreCase("APPROX_PERCENTILE");
+    }
+
+    private static class ApproxPercentilePatternMatcher
+            extends SqlParserDefaultVisitor
+    {
+        private final AstNode root;
+        private final ImmutableListMultimap.Builder<String, AstNode> firstArgToApproxPercentileNode = ImmutableListMultimap.builder();
+        private final ImmutableListMultimap.Builder<String, String> firstArgToPercentiles = ImmutableListMultimap.builder();
+
+        public ApproxPercentilePatternMatcher(AstNode root)
+        {
+            this.root = requireNonNull(root, "AST passed to pattern matcher was null");
+        }
+
+        public ApproxPercentilePatternMatcherResult matchPattern()
+        {
+            root.jjtAccept(this, null);
+            return new ApproxPercentilePatternMatcherResult(firstArgToApproxPercentileNode.build(), firstArgToPercentiles.build());
+        }
+
+        @Override
+        public void visit(FunctionCall node, Void data)
+        {
+            if (isApproxPercentileNode(node) && hasUnsignedLiteralSecondArg(node)) {
+                String firstArg = unparse(getNthArgument(node, 0).get()).trim();
+                String secondArg = unparse(getNthArgument(node, 1).get()).trim();
```

Comment:
You don't need to unparser the second argument?

Reply:
Agreed -- removed unparsing when we deal with secondArg.